src/libpmemobj/tx.c

```diff
@@ -668,6 +668,42 @@ tx_realloc_common(struct tx *tx, PMEMoid oid, size_t size, uint64_t type_num,
 	return new_obj;
 }
 
+/*
+ * tx_construct_prealloc -- add user preallocated buffer to the ulog
+ */
+static int
+tx_construct_prealloc(struct tx *tx, void *addr, size_t size, uint64_t type,
+		uint64_t outer_tx, size_t first_prealloc)
+{
+	if (tx->pop != pmemobj_pool_by_ptr(addr)) {
+		ERR("Allocation from different pool");
+		goto err;
+	}
+
+	switch (type) {
+	case TX_PARAM_PREALLOC_REDO:
+		if (operation_add_extend(tx->lane->external, addr, size))
+			goto err;
+		break;
+	case TX_PARAM_PREALLOC_UNDO:
+		if (outer_tx && first_prealloc)
+			/* free all logs except first */
+			operation_free_logs(tx->lane->undo);
```

Comment:
Why do you do this? If this is the first outer transaction, there won't be any logs to remove if I understand correctly.

Or is this only to remove the potentially allocated second ulog?